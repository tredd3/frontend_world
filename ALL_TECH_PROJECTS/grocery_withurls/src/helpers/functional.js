export const pipe = (...args) => x => args.reduce((acc, fn) => fn(acc), x);

export const range = length => [...Array(length).keys()];

export const pad = (length, padCharacter) => value => {
  if (String(value).length >= length) return String(value);
  return range(length - String(value).length).map(() => padCharacter).join('') + String(value);
};

const make2digit = pad(2, '0');

export const formatDate = dateLike => {
  const date = new Date(dateLike);
  return `${date.getFullYear()}-${make2digit(date.getMonth() + 1)}-${make2digit(date.getDate())}`;
};

export const objectToQueryString = obj => Object.entries(obj)
  .reduce((acc, [key, value]) => [...acc, `${encodeURIComponent(key)}=${encodeURIComponent(Array.isArray(value)
    ? value.join(',') : value)}`], []).join('&');

export const queryStringToObject = qs => {
  if (!qs) return {};
  return qs.split('&')
    .map(aQsPart => aQsPart.split('=').map(decodeURIComponent))
    .reduce((acc, [name, value]) => {
      const convertedValue = value.includes(',') ? value.split(',') : [value];
      return { ...acc, [name]: convertedValue };
    }, {});
};
