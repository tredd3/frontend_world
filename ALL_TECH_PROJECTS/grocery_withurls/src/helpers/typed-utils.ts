export const flatMap = <T, U>(arr: T[], fn: ((x: T) => U[])): U[] => (
  arr.reduce(
    (acc: U[], x: T) => [...acc, ...fn(x)],
    [] as U[]
  )
);
