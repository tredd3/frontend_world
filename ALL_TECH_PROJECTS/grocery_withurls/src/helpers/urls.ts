export const accountUrl = '/account';
export const cartUrl = '/cart';
export const notificationsUrl = '/notifications';
export const productUrl = (skuId: number | string) => `/products/${skuId}`;
export const productSimilarItemsUrl = (skuId: number | string) => `/products/${skuId}/similar-items`;
export const orderPlacedUrl = (orderId: number | string) => `/order-placed/${orderId}`;
export const ordersUrl = '/orders';
export const orderSearchUrl = '/orders/search';
export const orderDetailsUrl = (orderId: number | string) => `/orders/${orderId}`;
export const wishlistUrl = '/wishlist';
export const categoriesUrl = '/categories';
export const categoryUrl = (categoryId: number | string) => `/categories/${categoryId}`;
export const subCategoryUrl = (categoryId: number | string, subCategoryId: number | string) => `/categories/${categoryId}/${subCategoryId}`;
export const searchUrl = '/search';
export const searchResultsUrl = (searchText: string) => `/search/${searchText}`;
export const collectionUrl = (collectionId: number | string, type?: number | string, subtype?: number | string) => {
  let queryParams;
  if (type) {
    queryParams = new URLSearchParams();
    queryParams.append('type', String(type));

    if (subtype) {
      queryParams.append('subtype', String(subtype));
    }
  }
  return `/collections/${collectionId}${queryParams ? `?${queryParams.toString()}` : ''}`;
};
