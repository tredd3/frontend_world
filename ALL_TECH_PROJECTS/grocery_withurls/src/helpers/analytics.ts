import { getMobileOperatingSystem } from './utilities';
import { getUser } from '../services/user';
import { getCart, PaymentMethodId, paymentLabelById } from '../services/cart';

import { CheckoutData } from '../services/checkout';
import { CartPivot } from '../types';
import { flatMap } from './typed-utils';

let purchaseJourney: string;

export const setPurchaseJourney = (val: string) => {
  purchaseJourney = val;
};

export const getPurchaseJourney = () => purchaseJourney;

declare global {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    digitalData: Record<string, any>;
    _satellite: {
      track: (event: string) => void;
    };
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const track = async (event: string, data: Record<string, any>) => {
  if (typeof window === 'undefined') return;
  const { userId } = await getUser();
  const userOS = getMobileOperatingSystem();
  const platform = userOS === 2 ? 'Android App' : userOS === 3 ? 'iOS App' : 'Web';

  const digitalData = {
    link: {
      ...data.link
    },
    internal: {
      ...data.internal
    },
    page: {
      pageInfo: {
        appName: 'JioMart Web'
      },
      ...data.page
    },
    user: {
      platform,
      loginStatus: 'Logged in',
      martId: userId,
      jioCenter: '',
      ...data.user
    },
    product: [
      ...(data.product ? data.product : [])
    ],
    interPromotion: {
      ...data.interPromotion
    },
    ...(data.product ? data.filterSection : {}),
    ...(data.purchase ? data.purchase : {})
  };

  window.digitalData = {
    ...window.digitalData,
    ...digitalData
  };

  // eslint-disable-next-line no-underscore-dangle
  if (window._satellite) window._satellite.track(event);
};

type LinkPositions = 'Top' | 'Middle' | 'Bottom';

export const trackLink = (linkName: string, linkType: string, linkPosition: LinkPositions) => (
  track('linkTracking', { link: { linkName, linkType, linkPosition } })
);

export const trackPage = (pageName: string) => (
  track('pageLoad', { page: { pageName } })
);

type FilterSection = {
  filterSelection: boolean;
  sortSelection: boolean;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const trackProduct = (eventName: string, productData: Record<string, any>, filterSection?: FilterSection) => {
  const defaultFilterSection = {
    filterSelection: false,
    sortSelection: false
  };
  return track(eventName, {
    product: [productData],
    filterSection: (filterSection || defaultFilterSection)
  });
};

export const trackAPIShipments = (eventName: string, shipments: CheckoutData['Shipments'], purchase?: any) => {
  if (shipments && shipments.length) {
    track(eventName, {
      product: flatMap(shipments, shipment => shipment.Product)
        .map(product => ({
          productID: product.ProductSkuId,
          quantity: product.SelectedQuantity,
          price: product.SelectedQuantity * product.SP,
          productName: product.ProductName
        })),
      ...(purchase && { purchase })
    });
  }
};

type PurchaseArg = {
  paymentMethod: PaymentMethodId;
  paymentStatus?: 'Paid' | 'Unpaid';
  cart: CartPivot;
  failureReason?: string;
}

export const trackShipments = async (eventName: string, purchase?: PurchaseArg) => {
  const { shipments } = purchase?.cart || await getCart();
  if (!shipments || !shipments.length) return Promise.resolve();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let additionalInfo: Record<string, any> | null = null;

  if (purchase) {
    const { paymentMethod, cart, ...rest } = purchase;
    additionalInfo = {
      paymentMethod: paymentLabelById(paymentMethod),
      couponDiscount: cart.couponDiscountAmount,
      deliveryCharges: cart.deliveryCharge,
      totalOrderValue: cart.totalPayableAmount,
      transactionId: cart.cartId,
      ...rest
    };
  }

  return track(eventName, {
    product: flatMap(shipments, shipment => shipment.products)
      .map(product => ({
        productId: product.skuId,
        quantity: product.selectedQuantity,
        price: product.selectedQuantity * product.sp,
        productName: product.name
      })),
    ...(additionalInfo && additionalInfo)
  });
};
