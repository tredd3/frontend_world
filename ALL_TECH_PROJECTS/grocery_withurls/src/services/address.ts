import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser, updateUser } from './user';
import { APIBaseAddress, APIAddress, AddressPivot } from '../types';
import { pivotAPIAddress } from './helpers/pivots';

export type APIAddressRequest = APIBaseAddress & {
  addressId: number;
  addedDate: string;
};

// eslint-disable-next-line camelcase, @typescript-eslint/camelcase
export const DEPRECATED_convertAddressToAPIAddress = ({ id, addedDate, ...address }: APIAddress): APIAddressRequest => (
  { addressId: id, addedDate: addedDate.toDateString(), ...address }
);

type AreaType = { pincode: number};
export const isAreaServiceable = promiseMemoize(async ({ pincode }: AreaType) => {
  try {
    await makeApiCall<AreaType, {}>(apiRoutes.checkServiceability, { body: { pincode } });
    return true;
  } catch (e) {
    return false;
  }
}, { resolve: 'json' });

export const getAddresses = promiseMemoize(async (): Promise<AddressPivot[]> => {
  const { userId } = await getUser();
  const { addresses } = await makeApiCall<{ userId: number }, {addresses: APIAddressRequest[]}>(apiRoutes.getAddress, { body: { userId } });
  return addresses.map(pivotAPIAddress);
});

export const getAddressById = async (id: number): Promise<AddressPivot | undefined> => {
  const addresses = (await getAddresses()) || [];
  return addresses.find(address => address.id === id);
};

type UpdateAddressRequest = (APIAddressRequest | APIBaseAddress) & { userId?: number };

export const updateAddress = async (address: Omit<APIBaseAddress, 'address'> | APIAddress): Promise<APIAddress> => {
  const { id } = address as APIAddress;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { addedDate, ...remainingAddress } = address as APIAddress;
  const newAddress = await makeApiCall<UpdateAddressRequest, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      ...(id && { addressId: id }),
      ...({ userId: (await getUser()).userId }),
      ...remainingAddress,
      address: `${remainingAddress.cityName}, ${remainingAddress.pincode}`
    }
  });

  getAddresses.clear();
  return pivotAPIAddress(newAddress);
};

export const addAddress = (address: Omit<APIBaseAddress, 'address'>) => updateAddress(address);

export const deleteAddress = async (id: number) => {
  getAddresses.clear();
  return makeApiCall<{ addressId: number }, void>(
    apiRoutes.deleteAddress, { body: { addressId: id } }
  );
};

export const setDefaultAddress = async (id: number) => {
  const address = await getAddressById(id);
  if (!address) throw new Error('An address with this id doesn\'t exist');

  await makeApiCall<{ addressId: number; isDefault: boolean }, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      addressId: id,
      isDefault: true
    }
  });
  getAddresses.clear();

  await updateUser({
    lat: address.latitude,
    lng: address.longitude,
    pincode: address.pincode,
    storeId: address.storeId,
    storeName: address.storeName
  });

  return address;
};

export const getDefaultAddress = async () => {
  const addresses = await getAddresses();
  const defaultAddress = addresses.find(address => address.isDefault);
  if (defaultAddress) return defaultAddress;

  if (!addresses.length) return null;

  await setDefaultAddress(addresses[0].id);
  return addresses[0];
};

export const updateAddressWithKirana = async (addressId: number, { storeId, storeName }: { storeId: number; storeName: string }) => {
  const address = await getAddressById(addressId);

  if (!address) throw new Error('Couldn\'t find an address with this id');

  const updatedAddress = await updateAddress({ ...address, storeId, storeName });

  if (address.isDefault) {
    await updateUser({ storeId, storeName });
  }

  return updatedAddress;
};

export const deleteKiranaFromAddress = async (addressId: number) => updateAddressWithKirana(addressId, { storeId: 0, storeName: '' });

export const getStoreIdFromAddressId = async (addressId: number) => {
  const address = await getAddressById(addressId);

  if (!address) return null;

  return address.storeId || address.defaultStore || null;
};

export const getPincodeFromLatLng = async ({ lat, lng }: {lat: number; lng: number}): Promise<string> => {
  const geocoder = new window.google.maps.Geocoder();
  const pincode = new Promise<string>((resolve, reject) => {
    geocoder.geocode({ location: new window.google.maps.LatLng(lat, lng) }, (results, status) => {
      if (status === window.google.maps.GeocoderStatus.OK) {
        const len = results[0].address_components.length;
        resolve(results[0].address_components[len - 1].long_name);
      } else {
        reject();
      }
    });
  });
  return pincode;
};
