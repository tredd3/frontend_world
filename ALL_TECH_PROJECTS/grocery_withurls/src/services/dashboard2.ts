import promiseMemoize from 'promise-memoize';
import { APIDashboardProduct } from '../types/api';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { DashboardProductPivot } from '../types';
import { pivotDashboardProduct } from './helpers/pivots';
import { getUser } from './user';

type Banner = { id: number; imageUrl: string };
type BannerWithTitle = Banner & { title: string };
type WidgetOfType<T> = { type: T; subType: number };
type BannerWidget<T> = WidgetOfType<T> & { banners: Banner[] };

export type HorizontalScrollableIcons = WidgetOfType<1> & { banners: BannerWithTitle[] };
export type Pager = BannerWidget<3>;
export type ThinCarouselBanner = BannerWidget<4>;
export type BigSquareCarouselBanner = BannerWidget<5>;
export type SmallSquareCarouselBanner = BannerWidget<6> & { title: string }
export type HorizontalProducts = WidgetOfType<7> & {
  title: string;
  viewAllLabel: string;
  products: DashboardProductPivot[];
};
export type W4x = WidgetOfType<11> & { banners: BannerWithTitle[] };
export type W9x = WidgetOfType<12> & { banners: BannerWithTitle[] } & { title: string };
export type W4xWithBanner = BannerWidget<15> & { title: string };
export type W9xWithBanner = BannerWidget<16> & { title: string };

export type Widget =
  HorizontalScrollableIcons
  | Pager
  | ThinCarouselBanner
  | BigSquareCarouselBanner
  | SmallSquareCarouselBanner
  | HorizontalProducts
  | W4x
  | W9x
  | W4xWithBanner
  | W9xWithBanner

type DashboardRequestType = {
  pincode?: number;
  storeId?: number;
};

type APIBanner = {
  bannerId: number;
  imageUrl: string;
};

type APIMiscellaneous = {
  name: string;
  mscId: number;
  imageUrl: string;
}

type APIWidget = {
  type: Widget['type'];
  subType: number;
  title: string;
  deepLinkText: string;
  data: {
    banners: null | APIBanner[];
    products: null | APIDashboardProduct[];
    miscellaneous: null | APIMiscellaneous[];
  };
};

export type DashboardResponseType = {
  widgets: APIWidget[];
};

const widgetBase = (widget: APIWidget) => ({
  type: widget.type,
  subType: widget.subType
});

const apiBannerToBanner = (banner: APIBanner | APIMiscellaneous): Banner => ({
  id: (banner as APIBanner).bannerId || (banner as APIMiscellaneous).mscId,
  imageUrl: banner.imageUrl
});

const createBanners = ({ data: widgetData }: APIWidget): { banners: Banner[] } => ({
  banners: ((widgetData.banners || widgetData.miscellaneous || []) as (APIBanner | APIMiscellaneous)[])
    .map(apiBannerToBanner)
});

const pivotBanner = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBanners(widget)
});

const createBannersWithTitle = (widget: APIWidget): { banners: BannerWithTitle[] } => ({
  banners: ((widget.data.miscellaneous || [])as APIMiscellaneous[])
    .map(banner => ({
      ...apiBannerToBanner(banner),
      title: banner.name
    }))
});

const createProducts = (widget: APIWidget): { products: DashboardProductPivot[] } => ({
  products: widget.data.products!.map(pivotDashboardProduct)
});

const pivotBannersWithTitle = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBannersWithTitle(widget)
});

const pivotProducts = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createProducts(widget)
});

export const pivotDashboardResponse = ({ widgets }: DashboardResponseType) => (
  widgets
    .filter(w => w.data.banners || w.data.miscellaneous || w.data.products)
    .map(widget => {
      switch (widget.type) {
        case 1: return pivotBannersWithTitle(widget) as HorizontalScrollableIcons;
        case 3: return pivotBanner(widget) as Pager;
        case 4: return pivotBanner(widget) as ThinCarouselBanner;
        case 5: return pivotBanner(widget) as BigSquareCarouselBanner;
        case 6:
          return {
            ...widgetBase(widget),
            title: widget.title,
            ...createBanners(widget)
          } as SmallSquareCarouselBanner;
        case 7:
          return {
            title: widget.title,
            viewAllLabel: widget.deepLinkText,
            ...pivotProducts(widget)
          } as HorizontalProducts;
        case 11: return pivotBannersWithTitle(widget) as W4x;
        case 12:
          return {
            title: widget.title,
            ...pivotBannersWithTitle(widget)
          } as W9x;
        case 15: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W4xWithBanner;
        case 16: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W9xWithBanner;
        default: return null;
      }
    })
    .filter(<T>(x: T | null): x is T => x !== null)
);

export const getWidgetDefinitions = promiseMemoize(async () => {
  const { preferences: { pincode, storeId = 0 } } = await getUser();

  const apiWidgets = await makeApiCall<DashboardRequestType, DashboardResponseType>(apiRoutes.getDashboard, {
    body: { pincode, storeId }
  });

  return pivotDashboardResponse(apiWidgets);
}, { resolve: 'json' });
