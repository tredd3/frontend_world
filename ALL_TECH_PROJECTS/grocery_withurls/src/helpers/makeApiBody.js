import store from '../store';
import { getUser } from '../services/user';
import { getCart } from '../services/cart';

const getCartId = () => store.getState().cart.CartId || 0;
const getSelectedStoreId = async () => {
  const user = await getUser();
  const { preferences: { storeId } } = user;
  return storeId || undefined;
};

export const getUserId = () => store.getState().user.userId;

export const addCartApiBody = async ({
  skuId, Quantity = 1, ProductType = 0, deliveryModes, addressIdForCheckout
}) => {
  let apiBody = {};

  apiBody = {
    product: [{
      skuId, qty: Number(Quantity), productType: ProductType
    }]
  };

  // Here action is mandatory parameter to be passed. Because based on it backend internally calls checkout api.
  // action : 1 means checkout call
  // action : 0 means no checkout call
  // Also we need to pass other parameters like deliveryDate, timeSlot which needed for checkout api call at backend
  // when action is 1

  if (deliveryModes && Object.keys(deliveryModes).length !== 0) {
    const { date, timeSlot } = deliveryModes;
    apiBody = {
      ...apiBody, action: 1, deliveryDate: date, timeslot: timeSlot, customerAddressId: addressIdForCheckout
    };
  } else {
    apiBody = { ...apiBody, action: 0 };
  }

  const user = await getUser();
  if (user) {
    const { userId } = user;
    apiBody = { ...apiBody, userId };
  }

  const storeId = await getSelectedStoreId();
  if (storeId) {
    apiBody = { ...apiBody, storeId };
  }
  try {
    const cart = await (storeId ? getCart(storeId) : getCart());
    if (cart && cart.cartId !== 0) {
      const { cartId } = cart;
      apiBody = { ...apiBody, cartId };
    }
  } catch (e) {
    console.log(e.message);
  }

  return apiBody;
};

export const removeCartApiBody = ({ skuId, Quantity = 0, ProductType = 0 }) => ({
  cartId: getCartId(),
  skuId,
  qty: Quantity,
  type: ProductType
});

export const moveItemApiBody = ({
  skuId, Quantity = 0, ProductType = 0, moveType
}) => ({
  cartId: getCartId(),
  skuId,
  qty: Quantity,
  productType: ProductType,
  moveType
});

export const removeSaveLaterCartApiBody = ({
  skuId, Quantity = 0, ProductType = 0, moveType
}) => ({
  skuId,
  qty: Quantity,
  productType: ProductType,
  productCategoryType: moveType
});
