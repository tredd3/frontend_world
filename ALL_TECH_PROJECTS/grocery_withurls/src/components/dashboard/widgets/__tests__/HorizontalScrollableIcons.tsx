import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { mocked } from 'ts-jest/utils';
import HorizontalScrollableIconsWidget from '../HorizontalScrollableIcons';
import { pivotDashboardResponse, HorizontalScrollableIcons, DashboardResponseType } from '../../../../services/dashboard2';
import mockDB from '../../../../mocks/db.json';
import * as analytics from '../../../../helpers/analytics';

jest.mock('../../../../helpers/analytics');
jest.mock('../../../uiControls/HorizontalSlider', () => {
  const MockedHorizontalSlider: React.FC = ({ children }) => (
    <div>
      <div>HorizontalSlider Placeholder</div>
      {children}
    </div>
  );

  return {
    __esModule: true,
    default: MockedHorizontalSlider
  };
});

const { setPurchaseJourney, trackLink } = mocked(analytics);

const mockWidgetData = pivotDashboardResponse(mockDB.dashboard.Data as DashboardResponseType)
  .find(w => w.type === 1) as HorizontalScrollableIcons;

describe('HorizontalScrollableIconsWidget tests', () => {
  beforeEach(() => {
    trackLink.mockClear();
    setPurchaseJourney.mockClear();
  });

  it('should render the widget correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <HorizontalScrollableIconsWidget
          widgetData={mockWidgetData}
        />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should handle click action correctly for banner links', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <HorizontalScrollableIconsWidget
          widgetData={mockWidgetData}
        />
      </MemoryRouter>
    );

    mockWidgetData.banners.forEach((_banner, index) => {
      fireEvent.click(getByTestId(`banner-link-${index}`));
      expect(trackLink).toHaveBeenCalledTimes(index + 1);
      expect(trackLink.mock.calls[index]).toMatchSnapshot();
      expect(setPurchaseJourney).toHaveBeenCalledTimes(index + 1);
      expect(setPurchaseJourney.mock.calls[index]).toMatchSnapshot();
    });
  });
});
