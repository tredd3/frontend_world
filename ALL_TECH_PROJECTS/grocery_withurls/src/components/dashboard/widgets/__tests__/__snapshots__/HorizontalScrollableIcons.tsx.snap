// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 1`] = `
Array [
  "Shop By Category | Beauty & Hygiene",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 2`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 3`] = `
Array [
  "Shop By Category | Cleaning & Household Care",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 4`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 5`] = `
Array [
  "Shop By Category | Dairy, Bakery & Egg",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 6`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 7`] = `
Array [
  "Shop By Category | Foodgrains, Oils And Masala",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 8`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 9`] = `
Array [
  "Shop By Category | Snacks And Confectionary",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 10`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 11`] = `
Array [
  "Shop By Category | Health Drinks & Beverages",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 12`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 13`] = `
Array [
  "Shop By Category | Mom & Baby",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 14`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 15`] = `
Array [
  "Shop By Category | Instant And Ready Foods",
  "HomeScreenClicks",
  "Top",
]
`;

exports[`HorizontalScrollableIconsWidget tests should handle click action correctly for banner links 16`] = `
Array [
  "Shop By Category",
]
`;

exports[`HorizontalScrollableIconsWidget tests should render the widget correctly 1`] = `
<DocumentFragment>
  .emotion-3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  width: 76px;
  margin: 10px 5px;
  text-align: center;
}

.emotion-3:hover,
.emotion-3:focus,
.emotion-3:visited,
.emotion-3:active {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
}

.emotion-0 {
  width: 46px;
  height: 46px;
  border-radius: 50%;
  margin: 0px auto 5px;
}

.emotion-2 {
  width: 76px;
  font-size: 11px;
  color: #000;
  font-weight: 600;
}

.emotion-1 {
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
  overflow-y: hidden;
}

<div>
    <div>
      HorizontalSlider Placeholder
    </div>
    <a
      class="emotion-3"
      data-testid="banner-link-0"
      href="/categories/10003388"
    >
      <div
        class="emotion-0"
        data-testid="category0"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Beauty & Hygiene"
            src="https://preprod-static.jiomoney.com/static//category/10003388_Beauty-Hygiene_1569240984.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Beauty & Hygiene
        </div>
      </div>
    </a>
    <a
      class="emotion-3"
      data-testid="banner-link-1"
      href="/categories/10003394"
    >
      <div
        class="emotion-0"
        data-testid="category1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Cleaning & Household Care"
            src="https://preprod-static.jiomoney.com/static//category/10003394_Cleaning-HouseholdCare_1568986181.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Cleaning & Household Care
        </div>
      </div>
    </a>
    <a
      class="emotion-3"
      data-testid="banner-link-2"
      href="/categories/10003615"
    >
      <div
        class="emotion-0"
        data-testid="category2"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Dairy, Bakery & Egg"
            src="https://preprod-static.jiomoney.com/static//category/10003615_Dairy-Bakery-Egg_1568986212.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Dairy, Bakery & Egg
        </div>
      </div>
    </a>
    <a
      class="emotion-3"
      data-testid="banner-link-3"
      href="/categories/10003627"
    >
      <div
        class="emotion-0"
        data-testid="category3"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Foodgrains, Oils And Masala"
            src="https://preprod-static.jiomoney.com/static//category/10003627_Foodgrains-OilsAndMasala_1568986234.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Foodgrains, Oils And Masala
        </div>
      </div>
    </a>
    <a
      class="emotion-3"
      data-testid="banner-link-4"
      href="/categories/10003391"
    >
      <div
        class="emotion-0"
        data-testid="category4"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Snacks And Confectionary"
            src="https://preprod-static.jiomoney.com/static//category/10003391_SnacksAndConfectionary_1568986320.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Snacks And Confectionary
        </div>
      </div>
    </a>
    <a
      class="emotion-3"
      data-testid="banner-link-5"
      href="/categories/10003397"
    >
      <div
        class="emotion-0"
        data-testid="category5"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Health Drinks & Beverages"
            src="https://preprod-static.jiomoney.com/static//category/10003397_HealthDrinks-Beverages_1568986259.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Health Drinks & Beverages
        </div>
      </div>
    </a>
    <a
      class="emotion-3"
      data-testid="banner-link-6"
      href="/categories/10003400"
    >
      <div
        class="emotion-0"
        data-testid="category6"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Mom & Baby"
            src="https://preprod-static.jiomoney.com/static//category/10003400_Mom-Baby_1568986300.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Mom & Baby
        </div>
      </div>
    </a>
    <a
      class="emotion-3"
      data-testid="banner-link-7"
      href="/categories/10003549"
    >
      <div
        class="emotion-0"
        data-testid="category7"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 46px;"
        >
          <img
            alt="Instant And Ready Foods"
            src="https://preprod-static.jiomoney.com/static//category/10003549_InstantAndReadyFoods_1568986281.png?impolicy=resizeforwidth&w=46"
            width="46"
          />
        </span>
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        >
          Instant And Ready Foods
        </div>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
